File: app/src/main/java/no/nav/tilbakemeldingsmottak/rest/serviceklage/service/support/OpprettServiceklageRequestMapper.java
package no.nav.tilbakemeldingsmottak.rest.serviceklage.service.support;

import no.nav.tilbakemeldingsmottak.model.OpprettServiceklageRequest;
import no.nav.tilbakemeldingsmottak.model.OpprettServiceklageRequest.KlagetyperEnum;
import no.nav.tilbakemeldingsmottak.serviceklage.Serviceklage;
import org.apache.commons.lang3.StringUtils;
import org.springframework.stereotype.Component;

import java.time.LocalDateTime;
import java.util.List;
import java.util.stream.Collectors;

import static no.nav.tilbakemeldingsmottak.serviceklage.ServiceklageConstants.*;
import static org.apache.commons.lang3.StringUtils.isBlank;

@Component
public class OpprettServiceklageRequestMapper {

    public Serviceklage map(OpprettServiceklageRequest request, boolean innlogget) {
        LocalDateTime currentDateTime = LocalDateTime.now();
        return Serviceklage.builder()
                .opprettetDato(currentDateTime)
                .fremmetDato(currentDateTime.toLocalDate())
                .innsender(request.getPaaVegneAv().value)
                .klagenGjelderId(findKlagenGjelderId(request))
                .innlogget(innlogget)
                .klagetyper(mapKlagetype(request.getKlagetyper()))
                .klagetypeUtdypning(request.getKlagetyper().contains(KlagetyperEnum.ANNET) ? request.getKlagetypeUtdypning() : null)
                .gjelderSosialhjelp(request.getGjelderSosialhjelp() == null ? null : request.getGjelderSosialhjelp().value)
                .klagetekst(request.getKlagetekst())
                .svarmetode(mapSvarmetode(request.getOenskerAaKontaktes()))
                .svarmetodeUtdypning(mapSvarmetodeUtdypning(request.getOenskerAaKontaktes()))
                .kanal(KANAL_SERVICEKLAGESKJEMA_ANSWER)
                .enhetsnummerPaaklaget(!isBlank(request.getEnhetsnummerPaaklaget()) ? request.getEnhetsnummerPaaklaget() : null)
                .build();
    }

    private String mapKlagetype(List<KlagetyperEnum> klagetype) {
        return StringUtils.join(klagetype.stream().map(x -> x.value).collect(Collectors.toList()), ", ");
    }

    private String mapSvarmetode(Boolean oenskerAaKontaktes) {
        if (oenskerAaKontaktes == null) {
            return SVAR_IKKE_NOEDVENDIG_ANSWER;
        } else {
            return oenskerAaKontaktes ? null : SVAR_IKKE_NOEDVENDIG_ANSWER;
        }
    }

    private String mapSvarmetodeUtdypning(Boolean oenskerAaKontaktes) {
        if (oenskerAaKontaktes == null) {
            return INNMELDER_MANGLER_FULLMAKT_ANSWER;
        } else {
            return oenskerAaKontaktes ? null : BRUKER_IKKE_BEDT_OM_SVAR_ANSWER;
        }
    }

    private String findKlagenGjelderId(OpprettServiceklageRequest request) {
        String klagenGjelderId = null;
        switch (request.getPaaVegneAv()) {
            case PRIVATPERSON:
                klagenGjelderId = request.getInnmelder().getPersonnummer();
                break;
            case ANNEN_PERSON:
                klagenGjelderId = request.getPaaVegneAvPerson().getPersonnummer();
                break;
            case BEDRIFT:
                klagenGjelderId = request.getPaaVegneAvBedrift().getOrganisasjonsnummer();
                break;
        }
        return klagenGjelderId;
    }

}

File: app/src/main/kotlin/no/nav/tilbakemeldingsmottak/rest/serviceklage/service/support/OpprettServiceklageRequestMapper.kt
package no.nav.tilbakemeldingsmottak.rest.serviceklage.service.support

import no.nav.tilbakemeldingsmottak.domain.ServiceklageConstants.BRUKER_IKKE_BEDT_OM_SVAR_ANSWER
import no.nav.tilbakemeldingsmottak.domain.ServiceklageConstants.INNMELDER_MANGLER_FULLMAKT_ANSWER
import no.nav.tilbakemeldingsmottak.domain.ServiceklageConstants.KANAL_SERVICEKLAGESKJEMA_ANSWER
import no.nav.tilbakemeldingsmottak.domain.ServiceklageConstants.SVAR_IKKE_NOEDVENDIG_ANSWER
import no.nav.tilbakemeldingsmottak.domain.models.Serviceklage
import no.nav.tilbakemeldingsmottak.model.OpprettServiceklageKlagetype
import no.nav.tilbakemeldingsmottak.model.OpprettServiceklagePaaVegneAv.*
import no.nav.tilbakemeldingsmottak.model.OpprettServiceklageRequest
import org.springframework.stereotype.Component
import java.time.LocalDateTime

@Component
class OpprettServiceklageRequestMapper {
    fun map(request: OpprettServiceklageRequest, innlogget: Boolean): Serviceklage {
        val currentDateTime = LocalDateTime.now()
        return Serviceklage(
            opprettetDato = currentDateTime,
            fremmetDato = currentDateTime.toLocalDate(),
            innsender = request.paaVegneAv?.value,
            klagenGjelderId = findKlagenGjelderId(request),
            innlogget = innlogget,
            klagetyper = mapKlagetype(request.klagetyper),
            klagetypeUtdypning = mapKlagetypeUtdypning(request),
            gjelderSosialhjelp = request.gjelderSosialhjelp?.value,
            klagetekst = request.klagetekst,
            svarmetode = mapSvarmetode(request.oenskerAaKontaktes),
            svarmetodeUtdypning = mapSvarmetodeUtdypning(request.oenskerAaKontaktes),
            kanal = KANAL_SERVICEKLAGESKJEMA_ANSWER,
            enhetsnummerPaaklaget = mapEnhetsnummerPaaklaget(request),
        )
    }

    fun mapKlagetypeUtdypning(request: OpprettServiceklageRequest): String? {
        return if (request.klagetyper?.contains(OpprettServiceklageKlagetype.ANNET) == true) {
            request.klagetypeUtdypning
        } else null
    }

    private fun mapKlagetype(klagetype: List<OpprettServiceklageKlagetype>?): String {
        return klagetype?.joinToString(", ") { it.value } ?: ""
    }

    private fun mapSvarmetode(oenskerAaKontaktes: Boolean?): String? {
        return when (oenskerAaKontaktes) {
            null -> SVAR_IKKE_NOEDVENDIG_ANSWER
            true -> null
            false -> SVAR_IKKE_NOEDVENDIG_ANSWER
        }
    }

    private fun mapSvarmetodeUtdypning(oenskerAaKontaktes: Boolean?): String? {
        return when (oenskerAaKontaktes) {
            null -> INNMELDER_MANGLER_FULLMAKT_ANSWER
            true -> null
            false -> BRUKER_IKKE_BEDT_OM_SVAR_ANSWER
        }
    }

    fun mapEnhetsnummerPaaklaget(request: OpprettServiceklageRequest): String? {
        return if (request.enhetsnummerPaaklaget?.isNotBlank() == true) {
            request.enhetsnummerPaaklaget
        } else null
    }

    private fun findKlagenGjelderId(request: OpprettServiceklageRequest): String? {
        return when (request.paaVegneAv) {
            PRIVATPERSON -> request.innmelder?.personnummer
            ANNEN_PERSON -> request.paaVegneAvPerson?.personnummer
            BEDRIFT -> request.paaVegneAvBedrift?.organisasjonsnummer
            null -> throw IllegalArgumentException("PaaVegneAv kan ikke være null")
        }
    }
}
